File: sprint4/go/lesson10code2.go
func effectiveSolution(A int, x []int) map[[3]int]bool {
    history := make(map[int]bool)
    n := len(x)
    xSorted := make([]int, n)
    copy(xSorted, x)
    sort.Ints(xSorted)
    triples := make(map[[3]int]bool)
    for i := 0; i < n; i++ {
        for j := i + 1; j < n; j++ {
            target := A - xSorted[i] - xSorted[j]
            if history[target] {
                triples[[3]int{target, xSorted[i], xSorted[j]}] = true
            }
            history[xSorted[i]] = true
        }
    }
    return triples
}

File: sprint4/kotlin/lesson10code2.kt
fun effectiveSolution(A: Int, x: List<Int>): Set<Triple<Int, Int, Int>> {
    val history = HashSet<Int>()
    val n = x.size
    val xSorted = x.sorted()
    val triples = HashSet<Triple<Int, Int, Int>>()
    for (i in 0 until n) {
        for (j in i + 1 until n) {
            val target = A - xSorted[i] - xSorted[j]
            if (target in history) {
                // Заметим, что тут тройка уже отсортирована за счёт предварительной
                // сортировки всего массива.
                triples.add(Triple(target, xSorted[i], xSorted[j]))
            }
            history.add(xSorted[i])
        }
    }
    return triples
}